internal/rbd/controllerserver.go

```
@@ -357,6 +358,56 @@ func checkValidCreateVolumeRequest(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSna
 	return nil
 }
 
+func (cs *ControllerServer) ListSnapshots(
+	ctx context.Context,
+	req *csi.ListSnapshotsRequest,
+) (*csi.ListSnapshotsResponse, error) {
+	if err := cs.Driver.ValidateControllerServiceRequest(
+		csi.ControllerServiceCapability_RPC_LIST_SNAPSHOTS); err != nil {
+		log.ErrorLog(ctx, "invalid list snapshot req: %v", protosanitizer.StripSecrets(req))
+
+		return nil, err
+	}
+
+	mgr := NewManager(cs.Driver.GetInstanceID(), nil, req.GetSecrets())
+	defer mgr.Destroy(ctx)
+
+	snap, err := mgr.GetSnapshotByID(ctx, req.GetSnapshotId())
+	if err != nil {
+		return nil, status.Errorf(codes.NotFound, "failed to find snapshot with ID %q: %s", req.GetSnapshotId(), err.Error())
+	}
+	defer snap.Destroy(ctx)
+
+	// We need type casting here as SourceVolumeID is unset
+	// We fetch the source volume ID from the request and return
+	// that in the response
+	rbdSnap, ok := snap.(*rbdSnapshot)
```

The comment isn't explaining a good reason to typecast. `mgr.GetSnapshotByID()` should return a complete Snapshot object. The `SourceVolumeID` in the request is optional, so you can not use that, see the comment I left for CephFS.

---

@nixpanic Can this be addressed later when we have fix to store the SourceVolumeID in the omap?

---

`SourceVolumeID` won't be set when `req.GetSnapshotId()` is valid. So I do not think it will work as intended.

---

Main question here is, does anything break when the returned Snapshot object does not have its `SourceVolumeId` set? There is a reason why `snap.ToCSI()` fails because of it.

---

@nixpanic Yes you are correct, For the CreateSnapshot we need to have it set but not for List. it wont break as external snapshotter doesnt use SourceVolumeId for the snapshot it got as part of List operation, This PR is tested and working as expected.

---

From the [CSI Spec](https://github.com/container-storage-interface/spec/blob/master/spec.md#createsnapshot):

```
// Information about a specific snapshot.
message Snapshot {
  ...

  // Identity information for the source volume. Note that creating a
  // snapshot from a snapshot is not supported here so the source has to
  // be a volume. This field is REQUIRED.
  string source_volume_id = 3;
```

`.SourceVolumeId` is a REQUIRED attribute.

---

No, this RPC will only be called by external snapshotter as a part of its status update procedure for a snapshot.

The fields that are consumed are: `SizeBytes`, `CreationTime` and `ReadyToUse`

---

Yes from the Spec its required not used by external snapshotter.

---

Still sounds very dangerous to me. Have you guys checked that after getting and setting the updated fields in the VolumeSnapshot CR, the SourceVolumeId does not get wiped?

When an updated external-snapshotter is used, this will need to be verified again too.

---

Instead of doing the typecasting, maybe just hack `snapshot.ToCSI()` and do not return an error if `SourceVolumeID` is not set? That would be way cleaner.

---

I agree but it is used in other places as well and I'm not sure if suppressing the error is the right call. In case of `ListSnapshots` we know for sure that we won't ever be depending on `SourceVolumeID`

---

The error checking was added recently. As long as there are no further downstream-only changes to CreateSnapshot and VolumeGroupSnapshot calls, changing `ToCSI()` would be okay (definitely easier than typecasting and copying `ToCSI()` in this function).

---

Done

---

@Madhu-1 please see Niels' comment above wrt. https://github.com/red-hat-storage/ceph-csi/pull/538#discussion_r2115361017.

If we have an error in `ToCSI` the returned value will be `nil` and hence we cannot be used.